### PR TITLE
Moved utility methods from CheckUtil to MissingJavadocMethodCheck

### LINKED ISSUE #13596: Move CheckUtil#isGetterMethod and CheckUtil#isSetterMethod to MissingJavadocMethodCheck
Noticed at https://github.com/checkstyle/checkstyle/pull/13555, the only place we use these utility methods is in MissingJavadocMethodCheck, we should just move them there along with associated tests.

### PR DESCRIPTION
**Pull request by relentless-pursuit**
Resolves #13596

---

### Regression :-
---

report 1 :- https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/729c1ee_2023092722/reports/diff/index.html

---
Diff Regression config: https://gist.githubusercontent.com/relentless-pursuit/c6b58ae47d9f0419460753a8d5199535/raw/5cb0276c3f8ff2e86ac1b1942ebfc6493f7fc31c/my_checks.xml

### CODE REVIEW

**Comment by nrmancuso** [comment]
@relentless-pursuit please make CI happy, `mvn clean verify` should always be passing on your local before pushing changes to GitHub.

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocMethodCheckTest.java
Comment (nrmancuso) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocMethodCheckTest.java]: @romani can we just remove all artificial tests? We can allow pitest + regression testing to clean up code or find new test cases with real input.

Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocMethodCheckTest.java]: We can, but let's do this in separate PR, so we will do improvements step by step.

**Comment by romani** [comment]
@relentless-pursuit , please review CI failures, and address them

**Comment by relentless-pursuit** [comment]
> @relentless-pursuit , please review CI failures, and address them. https://app.circleci.com/pipelines/github/checkstyle/checkstyle/20353/workflows/b563a88b-fef6-4777-acf5-3dbf4905172c/jobs/365934/artifacts
> 
> https://github.com/checkstyle/checkstyle/actions/runs/5967016412/job/16187818505?pr=13607#step:7:347 We lost pitest test coverage.
> 
> Please look at #13555 On how @Kevin222004 , did regression testing to catch test cases to cover mutation.

I am new to mutation testing. I went through the logs and tried to redesign my test cases and test input file. But, i assume i am not there. though, running locally `mvn -P pitest-javadoc clean verify` i found build success. is it the command to run mutation tests locally?

**Comment by nrmancuso** [comment]
@relentless-pursuit please read through https://github.com/checkstyle/checkstyle/wiki/How-to-Generate-and-Analyze-Pitest-Reports

**Comment by relentless-pursuit** [comment]
> @relentless-pursuit please read through https://github.com/checkstyle/checkstyle/wiki/How-to-Generate-and-Analyze-Pitest-Reports. Though, I have seen the mutations using the CI-logs, but I am struggling with report generation.

I tried doing pitest report generation in my local but it timed out despite making changes to the timeout constant in pom.xml. i also tried making a report generation but it is failing strangely, while i had done earlier successfully. 

i have a doubt. when using the groovy command via single mode, i am being able to generate the report successfully. but, when doing it via diff mode, i am facing lot many violations. 

`There are 279065 errors reported by Checkstyle 10.12.3-SNAPSHOT with /Users/piyushkumarsadangi/Development/newPR/checkstyle/src/main/resources/sun_checks.xml ruleset.`

I am using the following groovy command:
```
groovy diff.groovy \
--localGitRepo /Users/piyushkumarsadangi/Development/newPR/checkstyle \
--baseBranch temp_upstream_master \
--baseConfig /Users/piyushkumarsadangi/Development/newPR/checkstyle/src/main/resources/sun_checks.xml \
--patchBranch master \
--patchConfig /Users/piyushkumarsadangi/Development/contribution/checkstyle-tester/my_check.xml \
--listOfProjects /Users/piyushkumarsadangi/Development/contribution/checkstyle-tester/projects-to-test-on.properties
```
`temp_upstream_master` is tracking `upstream/master`

**Comment by rnveach** [comment]
> --baseConfig /Users/piyushkumarsadangi/Development/newPR/checkstyle/src/main/resources/sun_checks.xml

Don't use sun config. Both configs should be `my_check.xml` and you need to modify the config to suit the regression needs. Please read the documentation associated to it and let us know if you are not understanding something.

**Comment by relentless-pursuit** [comment]
> > @relentless-pursuit please read through https://github.com/checkstyle/checkstyle/wiki/How-to-Generate-and-Analyze-Pitest-Reports. Though, I have seen the mutations using the CI-logs, but I am struggling with report generation.
> 
> I tried doing pitest report generation in my local but it timed out despite making changes to the timeout constant in pom.xml. i also tried making a report generation but it is failing strangely, while i had done earlier successfully.
> 
> i have a doubt. when using the groovy command via single mode, i am being able to generate the report successfully. but, when doing it via diff mode, i am facing lot many violations.
> 
> `There are 279065 errors reported by Checkstyle 10.12.3-SNAPSHOT with /Users/piyushkumarsadangi/Development/newPR/checkstyle/src/main/resources/sun_checks.xml ruleset.`
> 
> I am using the following groovy command:
> 
> ```
> groovy diff.groovy \
> --localGitRepo /Users/piyushkumarsadangi/Development/newPR/checkstyle \
> --baseBranch temp_upstream_master \
> --baseConfig /Users/piyushkumarsadangi/Development/newPR/checkstyle/src/main/resources/sun_checks.xml \
> --patchBranch master \
> --patchConfig /Users/piyushkumarsadangi/Development/contribution/checkstyle-tester/my_check.xml \
> --listOfProjects /Users/piyushkumarsadangi/Development/contribution/checkstyle-tester/projects-to-test-on.properties
> ```
> 
> `temp_upstream_master` is tracking `upstream/master`



> > --baseConfig /Users/piyushkumarsadangi/Development/newPR/checkstyle/src/main/resources/sun_checks.xml
> 
> Don't use sun config. Both configs should be `my_check.xml` and you need to modify the config to suit the regression needs. Please read the documentation associated to it and let us know if you are not understanding something.

Thank you for your response. I have been able to generate the mutations in my local. Ideally, i was terminating the process assuming something wrong when it said timeout. but, when i kept the process running for long, i did get the list. i read the documentation. and, a reference PR mentioned. I am yet to implement your suggesstion for regression report. Here is my mutations. 

```
New surviving mutation(s) found:

Source File: "MissingJavadocMethodCheck.java"
Class: "com.puppycrawl.tools.checkstyle.checks.javadoc.MissingJavadocMethodCheck"
Method: "isGetterMethod"
Line Contents: "if (matchesGetterFormat &amp;&amp; noParams) {"
Mutator: "org.pitest.mutationtest.engine.gregor.mutators.RemoveConditionalMutator_EQUAL_IF"
Description: "removed conditional - replaced equality check with true"
Line Number: 383

Source File: "MissingJavadocMethodCheck.java"
Class: "com.puppycrawl.tools.checkstyle.checks.javadoc.MissingJavadocMethodCheck"
Method: "isSetterMethod"
Line Contents: "&amp;&amp; ast.getChildCount() == SETTER_GETTER_MAX_CHILDREN) {"
Mutator: "org.pitest.mutationtest.engine.gregor.mutators.RemoveConditionalMutator_EQUAL_IF"
Description: "removed conditional - replaced equality check with true"
Line Number: 411

Source File: "MissingJavadocMethodCheck.java"
Class: "com.puppycrawl.tools.checkstyle.checks.javadoc.MissingJavadocMethodCheck"
Method: "isSetterMethod"
Line Contents: "if (matchesSetterFormat &amp;&amp; singleParam) {"
Mutator: "org.pitest.mutationtest.engine.gregor.mutators.RemoveConditionalMutator_EQUAL_IF"
Description: "removed conditional - replaced equality check with true"
Line Number: 419
```
I can see in the documentation talking about Generating a Behavior Diff Report via a new PR to discuss the mutated code. Is that the next step? Can you give me a outline on what to do next?

**Comment by relentless-pursuit** [comment]
Github, generate report

**Comment by relentless-pursuit** [comment]
Github, generate report

**Comment by relentless-pursuit** [comment]
Github, generate report

**Comment by relentless-pursuit** [comment]
Github, generate report

**Comment by relentless-pursuit** [comment]
Github, generate report

**Comment by relentless-pursuit** [comment]
> @relentless-pursuit , please review CI failures, and address them. https://app.circleci.com/pipelines/github/checkstyle/checkstyle/20353/workflows/b563a88b-fef6-4777-acf5-3dbf4905172c/jobs/365934/artifacts
> 
> https://github.com/checkstyle/checkstyle/actions/runs/5967016412/job/16187818505?pr=13607#step:7:347 We lost pitest test coverage.
> 
> Please look at #13555 On how @Kevin222004 , did regression testing to catch test cases to cover mutation.

do the reports look fine?

**Comment by nrmancuso** [comment]
> do the reports look fine?

@relentless-pursuit 

Please do:
1. Use all check modules from MissingJavadocMethodCheck. It looks like you just copy pasted the existing config from the contribution repo without adding any modules for this check.
2. Remove usage of projects file, this is really only for special cases. We have a default projects file that is used.

You must use MissingJavadocMethodCheck modules in the config.

**Comment by relentless-pursuit** [comment]
Github, generate report

**Comment by relentless-pursuit** [comment]
Github, generate report

**Comment by relentless-pursuit** [comment]
Github, generate report

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocMethodCheckTest.java
Comment (nrmancuso) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocMethodCheckTest.java:35]: @relentless-pursuit looks like we missed migration of some tests, this is why pitest is failing:
https://github.com/checkstyle/checkstyle/blob/a168d97509e57bf78290f8fd790daf7f43c73030/src/test/java/com/puppycrawl/tools/checkstyle/utils/CheckUtilTest.java#L425-L446.

Please do:
1. Migrate all MissingJavadocMethod tests in CheckUtilTest to this class.
2. Migrate all inputs to corresponding package.

We should see a removal of test inputs from CheckUtil input package and addition of new inputs in MissingJavadocMethod package.

Comment (relentless-pursuit) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocMethodCheckTest.java:35]: done. i moved `InputCheckUtil9.java` and renamed it to match the convention. however, `InputCheckUtil1.java` is still being used by another method in `CheckUtilTest.java`. So, I created another file rather moving it. 

On a side note, what does it take to become one of the maintainers of checkstyle?

Comment (relentless-pursuit) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocMethodCheckTest.java:35]: Any idea on how to tackle the mutations because I assume i migrated all CheckUtilTests related to MissingJavadocMethod tests?
New surviving mutation(s) found:

```
Source File: "MissingJavadocMethodCheck.java"
Class: "com.puppycrawl.tools.checkstyle.checks.javadoc.MissingJavadocMethodCheck"
Method: "isSetterMethod"
Line Contents: "if (matchesSetterFormat &amp;&amp; singleParam) {"
Mutator: "org.pitest.mutationtest.engine.gregor.mutators.RemoveConditionalMutator_EQUAL_IF"
Description: "removed conditional - replaced equality check with true"
Line Number: 419
```

Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocMethodCheckTest.java:35]: >On a side note, what does it take to become one of the maintainers of checkstyle?

Just keep sending PRs with fixes for tickets.
While you are contributing you are maintainer. Your work will be visible at GitHub insights and release notes.

Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocMethodCheckTest.java:35]: Line in question is
https://github.com/checkstyle/checkstyle/pull/13607/files#diff-b4400c927ed8fca06e707dc1bc59acd1c05dba7eaaccfce7fb77956965760bcfR419

Pitest claims that changing this expression to "true" makes code to work the same ( non of test failing).
We need to find Input file content that can fail a test if such mutation is present.

Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocMethodCheckTest.java:35]: > matchesSetterFormat &amp;&amp; singleParam

Be careful which conditional it is changing. It could be changing `matchesSetterFormat` or `singleParam` and not the whole expression. You should confirm first before it turns out you are going down the wrong path.

**Review by nrmancuso — CHANGES_REQUESTED** [review]
@relentless-pursuit please point to where we migrated this test case: https://github.com/checkstyle/checkstyle/blob/a6132c35974aa2872f6f0900ef9ea300d247d566/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/utils/checkutil/InputCheckUtil1.java#L11

All test cases must be migrated with exactly the same input/config.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocMethodCheckTest.java
Comment (relentless-pursuit) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocMethodCheckTest.java:35]: @romani, @rnveach thank you for such kind words and insights. 

i have been trying to write test cases to kill the mutation. 

for the recent push, i addded new assertions and they are correctly detecting mutations in my local, as they fail when the mutation is present but pass when the original condition is used. 

The first assertion which i ommitted (metioned below), passes in both the mutated and non-mutated cases. After omitting it, and running pitest locally, i didn't see any surviving mutations in suppression file. 
```
 assertWithMessage("Invalid result: AST provided is setter method")
                .that(MissingJavadocMethodCheck.isSetterMethod(setterMethod))
                .isTrue(); 
```

However, the above lines ensured to check a method which is a proper setter body. So, i have tried and spent a lot of time to tweak input file to kill mutation. Shall I proceed in some other direction? Or, suppressing the mutation can be a rationale choice?

**Comment by relentless-pursuit** [comment]
> @relentless-pursuit please point to where we migrated this test case:
> 
> https://github.com/checkstyle/checkstyle/blob/a6132c35974aa2872f6f0900ef9ea300d247d566/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/utils/checkutil/InputCheckUtil1.java#L11
> 
> All test cases must be migrated with exactly the same input/config.

InputCheckUtil1.java is being used in CheckUtilTest not related to MissingJavadocMethodCheck. 
```
@Test
    public void testEqualsAvoidNullCheck() throws Exception {

        final String[] expected = {
            "14:28: " + getCheckMessage(EqualsAvoidNullCheck.class, MSG_EQUALS_AVOID_NULL),
            "21:17: " + getCheckMessage(EqualsAvoidNullCheck.class, MSG_EQUALS_AVOID_NULL),
        };
        verifyWithInlineConfigParser(
                getPath("InputCheckUtil1.java"), expected);
    }
```

Hence, I created a separate file. Are you asking me to have the same names of the new files i created?

**Comment by nrmancuso** [comment]
@relentless-pursuit you are referencing the wrong test, this is the one that it looks like we are missing: https://github.com/nrmancuso/checkstyle/blob/a168d97509e57bf78290f8fd790daf7f43c73030/src/test/java/com/puppycrawl/tools/checkstyle/utils/CheckUtilTest.java#L440 attention to `non-compilable`

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocMethodCheckTest.java
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocMethodCheckTest.java:35]: >suppressing the mutation can be a rationale choice?

We do now a lot of efforts to cleanup all suppressions. So we do not allow extension of them.

**Comment by relentless-pursuit** [comment]
Github, generate report

**Comment by relentless-pursuit** [comment]
> @relentless-pursuit you are referencing the wrong test, this is the one that it looks like we are missing: https://github.com/nrmancuso/checkstyle/blob/a168d97509e57bf78290f8fd790daf7f43c73030/src/test/java/com/puppycrawl/tools/checkstyle/utils/CheckUtilTest.java#L440 attention to `non-compilable`

thank you for the insights. when i searched in my IDE, i didn't show both the input files with the same name. hence, i was confused and proceeded in a different direction. i assume the ci looks happy. can you review it again?

also, what attributes to constant failure of ci/circleci: run-inspections?

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocMethodCheckTest.java
Comment (nrmancuso) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocMethodCheckTest.java:35]: @relentless-pursuit to be clear: we should have no need to add more tests or pitest suppressions. We previously had no surviving pitest mutations in this code. If we migrate all tests and test inputs, we should have none now. We are still missing tests and test cases in the migration effort. All test methods and test inputs should be moved to the correct path, verbatim.

**Comment by nrmancuso** [comment]
> also, what attributes to constant failure of ci/circleci: run-inspections?

Click into failing jobs, read output, visit “ARTIFACTS” tab, read report.

**Comment by relentless-pursuit** [comment]
> > also, what attributes to constant failure of ci/circleci: run-inspections?
> 
> Click into failing job, read output, visit “ARTIFACTS” tab, read report.

done

File: src/test/java/com/puppycrawl/tools/checkstyle/utils/CheckUtilTest.java
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/utils/CheckUtilTest.java:419]: make like:
`public static DetailAST getNode(File file, int type)`

**Review by romani — CHANGES_REQUESTED** [review]
last item:

File: src/test/java/com/puppycrawl/tools/checkstyle/utils/CheckUtilTest.java
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/utils/CheckUtilTest.java]: Not done.

@relentless-pursuit , please do not request review, and just reply to each and every review comment as "done" to confirm you saw it and pushed it

Comment (relentless-pursuit) [reply src/test/java/com/puppycrawl/tools/checkstyle/utils/CheckUtilTest.java]: Hi @romani. I will keep that in mind. Maybe, I was in a rush to merge the PR. 

Anyways, I assume you asked me to change the file name from `getNodeFromSpecificFile(int type, File file)` to `getNode(File file, int type)`. I have done that possibly. Please do let me know if i understood you right. And, the recent push is available for your review.

**Review by romani — APPROVED** [review]
Ok to merge

File: src/test/java/com/puppycrawl/tools/checkstyle/utils/CheckUtilTest.java
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/utils/CheckUtilTest.java]: All good. Thank a lot for fix.

**Comment by relentless-pursuit** [comment]
rebased

**Comment by romani** [comment]
There is a conflict please resolve

**Comment by relentless-pursuit** [comment]
> There is a conflict please resolve

Done. Resolved conflicts and Rebased again. Can we merge it?

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Blocked until we have resolution at https://github.com/checkstyle/checkstyle/pull/13407

**Comment by romani** [comment]
Please recheck my answer, better to not block and proceed.

**Comment by relentless-pursuit** [comment]
> Blocked until we have resolution at #13407

It seems the blocking PR is merged. And, as i can decipher from the comments, that we could proceed with this issue and improve later on handling such issues. I have rebased. Can i expect a merge?

**Comment by romani** [comment]
@relentless-pursuit , no need to rebase.
@nrmancuso , will come back to review soon.

**Review by nrmancuso — APPROVED** [review]

**Comment by nrmancuso** [comment]
>And, on a conceptual level, i wanted to understand something. Initially, when i migrated the setter and getter to MissingJavadocMethodCheckTest.java class, it showed me a pitest failure. But, when i followed your guidelines of migrating some other methods remanining in CheckUtilTest.java, the pitest failure dissapeared. Is it because, the methods like missingJavadoc() and testSetterGetterOn() were related? I am asking this because i spent a lot of time trying to kill the mutation while the actual logic was to migrate above methods. If i could understand a pattern here, maybe i can be quick in my furture PRs.

@relentless-pursuit here is what was happening:
1. We moved utility methods to check
2. We moved some (but not all) tests to check test
3. Pitest mutated (changed code) in the migrated methods
4. Now, we did not have tests that would fail for the way that pitest changed the code
5. Our pitest system failed your build

As you saw, once we put ALL the previous tests in place, we now had tests that would fail when pitest changed our code, thus "killing" the mutations.

I would suggest to take up a pitest issue to better understand how this all works. Also, the pitest website is a good resource: https://pitest.org/.

@Kevin222004 can you suggest a good beginner mutation for @relentless-pursuit to take up?

**Comment by relentless-pursuit** [comment]
> > And, on a conceptual level, i wanted to understand something. Initially, when i migrated the setter and getter to MissingJavadocMethodCheckTest.java class, it showed me a pitest failure. But, when i followed your guidelines of migrating some other methods remanining in CheckUtilTest.java, the pitest failure dissapeared. Is it because, the methods like missingJavadoc() and testSetterGetterOn() were related? I am asking this because i spent a lot of time trying to kill the mutation while the actual logic was to migrate above methods. If i could understand a pattern here, maybe i can be quick in my furture PRs.
> 
> @relentless-pursuit here is what was happening:
> 
> 1. We moved utility methods to check
> 2. We moved some (but not all) tests to check test
> 3. Pitest mutated (changed code) in the migrated methods
> 4. Now, we did not have tests that would fail for the way that pitest changed the code
> 5. Our pitest system failed your build
> 
> As you saw, once we put ALL the previous tests in place, we now had tests that would fail when pitest changed our code, thus "killing" the mutations.
> 
> I would suggest to take up a pitest issue to better understand how this all works. Also, the pitest website is a good resource: https://pitest.org/.
> 
> @Kevin222004 can you suggest a good beginner mutation for @relentless-pursuit to take up?

thank you @nrmancuso for the clarification. sure, would love to work on a pitest issue.

**Comment by Kevin222004** [comment]
@nrmancuso The majority of Mutations related to checks are almost finished few are left but they are complex. I think 
@relentless-pursuit you should try to figure out the mutation of `JavadocMethodCheck` 
https://github.com/checkstyle/checkstyle/blob/c7b3fd90624414dffd86948340eb78f714c8ba83/config/pitest-suppressions/pitest-javadoc-suppressions.xml#L93-L262
and `JavadocStyleCheck` https://github.com/checkstyle/checkstyle/blob/c7b3fd90624414dffd86948340eb78f714c8ba83/config/pitest-suppressions/pitest-javadoc-suppressions.xml#L462-L487  This might not be too complex to solve.

**Comment by relentless-pursuit** [comment]
> @nrmancuso The majority of Mutations related to checks are almost finished few are left but they are complex. I think @relentless-pursuit you should try to figure out the mutation of `JavadocMethodCheck` or `JavadocStyleCheck` This might not be too complex to solve.

Thanks @Kevin222004

**Pull request merged**
